contract_sale_generation/models/contract_line.py

```diff
@@ -1,41 +1,41 @@
 # Copyright (C) 2020 Open Source Integrators
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from odoo import api, models
+from odoo import models
 
 
 class ContractLine(models.Model):
     _inherit = "contract.line"
 
-    @api.multi
-    def _prepare_sale_line(self, order_id=False, sale_values=False):
-        self.ensure_one()
-        dates = self._get_period_to_invoice(
-            self.last_date_invoiced, self.recurring_next_date
-        )
+    def _prepare_sale_line_vals(self, dates, order_id=False):
         sale_line_vals = {
             "product_id": self.product_id.id,
             "product_uom_qty": self._get_quantity_to_invoice(*dates),
-            "uom_id": self.uom_id.id,
+            "product_uom": self.uom_id.id,
             "discount": self.discount,
             "contract_line_id": self.id,
             "display_type": self.display_type,
         }
         if order_id:
             sale_line_vals["order_id"] = order_id.id
+        return sale_line_vals
+
+    def _prepare_sale_line(self, order_id=False, sale_values=False):
+        self.ensure_one()
+        dates = self._get_period_to_invoice(
+            self.last_date_invoiced, self.recurring_next_date
+        )
+        sale_line_vals = self._prepare_sale_line_vals(dates, order_id)
+
         order_line = (
             self.env["sale.order.line"]
-            .with_context(
-                force_company=self.contract_id.company_id.id,
-            )
+            .with_company(self.contract_id.company_id.id)
```

Comment:
`with_company` expects a recordset, isn't it?

Reply:
That does not have importance:

![image](https://user-images.githubusercontent.com/19529533/151312265-0932a521-7418-498d-aed3-7342305a9ef5.png)